pkg/controllers/clusterresourceplacement/placement_status.go

```diff
@@ -342,3 +342,360 @@ func buildFailedResourcePlacements(work *fleetv1beta1.Work) (isPending bool, res
+			continue
+		}
+
+		// We don't check the bindings[i].Spec.ResourceSnapshotName != latestResourceSnapshot.Name here.
+		// The existing conditions are needed when building the new ones.
+		if bindings[i].Spec.SchedulingPolicySnapshotName != latestSchedulingPolicySnapshot.Name {
+			continue
+		}
+		res[bindings[i].Spec.TargetCluster] = &bindings[i]
+	}
+	return res, nil
+}
+
+// setResourcePlacementStatusPerCluster sets the resource related fields for each cluster.
+// It returns an array which records the status for each resource condition.
+// The resource condition order (index) is defined as const:
+// const (
+//
+//	RolloutStartedCondition resourceCondition = iota
+//	OverriddenCondition
+//	WorkCreatedCondition
+//	AppliedCondition
+//	AvailableCondition
+//	TotalCondition
+//
+// )
+func (r *Reconciler) setResourcePlacementStatusPerCluster(ctx context.Context,
+	crp *fleetv1beta1.ClusterResourcePlacement, latestResourceSnapshot *fleetv1beta1.ClusterResourceSnapshot, binding *fleetv1beta1.ClusterResourceBinding, status *fleetv1beta1.ResourcePlacementStatus) ([]metav1.ConditionStatus, error) {
+	if binding == nil {
+		meta.SetStatusCondition(&status.Conditions, RolloutStartedCondition.UnknownResourceConditionPerCluster(crp.Generation))
+		return []metav1.ConditionStatus{metav1.ConditionUnknown}, nil
+	}
+
+	res := make([]metav1.ConditionStatus, 0, TotalCondition)
+	// There are few cases:
+	// * if the resourceSnapshotName is not equal,
+	//     1. the status is false, it means the rollout is stuck.
+	//     2. otherwise, the rollout controller has not processed it yet.
+	// * if the resourceSnapshotName is equal,
+	//     just return the corresponding status.
+	if binding.Spec.ResourceSnapshotName == latestResourceSnapshot.Name {
+		for i := RolloutStartedCondition; i < TotalCondition; i++ {
+			bindingCond := binding.GetCondition(string(i.ResourceBindingConditionType()))
+			if !condition.IsConditionStatusTrue(bindingCond, binding.Generation) &&
+				!condition.IsConditionStatusFalse(bindingCond, binding.Generation) {
+				meta.SetStatusCondition(&status.Conditions, i.UnknownResourceConditionPerCluster(crp.Generation))
+				res = append(res, metav1.ConditionUnknown)
+				break
+			}
+			cond := metav1.Condition{
+				Type:               string(i.ResourcePlacementConditionType()),
+				Status:             bindingCond.Status,
+				ObservedGeneration: crp.Generation,
+				Reason:             bindingCond.Reason,
+				Message:            bindingCond.Message,
+			}
+
+			switch i {
+			case RolloutStartedCondition:
+				if bindingCond.Status == metav1.ConditionTrue {
+					status.ApplicableResourceOverrides = binding.Spec.ResourceOverrideSnapshots
+					status.ApplicableClusterResourceOverrides = binding.Spec.ClusterResourceOverrideSnapshots
+				}
+			case AppliedCondition, AvailableCondition:
+				if bindingCond.Status == metav1.ConditionFalse {
+					if err := r.setFailedPlacementsPerCluster(ctx, crp, binding, status); err != nil {
```

Comment:
IIRC, we will try to create the failedPlacements if the binding condition is false but the real condition is coming from work. At line 630, we will not create any failure if the work conditions are not current. Is this going to lead to any inconsistent status?